extensions-core/kinesis-indexing-service/src/main/java/org/apache/druid/indexing/kinesis/KinesisIndexTaskIOConfig.java

```diff
@@ -202,31 +215,55 @@ public String getEndpoint()
     return endpoint;
   }
 
-  @JsonProperty
-  public int getRecordsPerFetch()
+  @Nullable
+  @JsonProperty("recordsPerFetch")
+  @JsonInclude(JsonInclude.Include.NON_NULL)
+  public Integer getRecordsPerFetchConfigured()
   {
     return recordsPerFetch;
   }
 
+  public int getRecordsPerFetchOrDefault(final long maxHeapSize, final int fetchThreads)
+  {
+    if (recordsPerFetch != null) {
+      return recordsPerFetch;
+    } else {
+      final long memoryToUse = Math.min(
+          MAX_RECORD_FETCH_MEMORY,
+          (long) (maxHeapSize * RECORD_FETCH_MEMORY_MAX_HEAP_FRACTION)
+      );
+
+      final int assumedRecordSize = deaggregate
+                                    ? KinesisIndexTaskTuningConfig.ASSUMED_RECORD_SIZE_AGGREGATE
+                                    : KinesisIndexTaskTuningConfig.ASSUMED_RECORD_SIZE;
+
+      return Ints.checkedCast(Math.max(1, memoryToUse / assumedRecordSize / fetchThreads));
```

Comment:
Am I right to assume that for a 1GB-heap task with deaggregate=true and 1 available processor, the default recordsPerFetch from this calculation is 25?
```
memoryToUse = min(100_000_000, 1_000_000_000 * 0.05) = 50_000_000
assumedRecordSize = 1_000_000
recordsPerFetch = 50_000_000 / 1_000_000 / 2 = 25
```

If we carry the assumption that each record is 1MB, I think this result conflicts with the maximum size of data that GetRecords can return, which is 10 MB.

Reply:
My understanding of the Kinesis API is that in that case, the GetRecords API will return fewer records, using our limit as a cap. That would be OK if so; nothing wrong with getting less than this from the API.

Reply:
Ah, true